### PR TITLE
Feature/alternative template generation

### PR DESCRIPTION
**Pull request by RQuadling**
Hopefully, the final time I'm going to be making a pull request for this feature. So so sorry.

This PR replaces PR #444.

Rebased to 0.4.x-dev

### CODE REVIEW

File: src/Phinx/Console/Command/Create.php [hunk 1]
```diff
@@ -131,7 +183,21 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        }
+
         $output->writeln('<info>using migration base class</info> ' . $classes['$useClassName']);
+
+        if (!empty($altTemplate)) {
+            $output->writeln('<info>using alternative template</info> ' . realpath($altTemplate));
+        } elseif (!empty($creationClassName)) {
+            $output->writeln('<info>using template creation class</info> ' . $creationClassName);
+        } else {
+            $output->writeln('<info>using template</info> ' . realpath(dirname(__FILE__) . self::DEFAULT_MIGRATION_TEMPLATE));
```
Comment (shadowhand) [inline src/Phinx/Console/Command/Create.php]: I don't think this is necessary, it doesn't exist in the current code.

File: src/Phinx/Console/Command/Create.php [hunk 2]
```diff
@@ -131,7 +183,21 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
 
+        // Do we need to do the post creation call to the creation class?
+        if ($creationClassName) {
+            $creationClass->postMigrationCreation($filePath, $className, $this->getConfig()->getMigrationBaseClassName());
+        }
+
         $output->writeln('<info>using migration base class</info> ' . $classes['$useClassName']);
+
+        if (!empty($altTemplate)) {
+            $output->writeln('<info>using alternative template</info> ' . realpath($altTemplate));
```
Comment (shadowhand) [inline src/Phinx/Console/Command/Create.php]: I wouldn't use `realpath` here, as it is not really necessary to show the full path. Reflecting the option the user entered should be enough.

File: src/Phinx/Migration/CreationInterface.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * Phinx
+ *
+ * (The MIT license)
+ * Copyright (c) 2014 Rob Morgan
```
Comment (shadowhand) [inline src/Phinx/Migration/CreationInterface.php]: 2015 :)

File: src/Phinx/Console/Command/Create.php [hunk 4]
```diff
@@ -106,15 +119,54 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
             ));
         }
 
-        // load the migration template
-        $contents = $this->getMigrationTemplate();
+        // Get the alternative template and static class options, but only allow one of them.
+        $altTemplate = $input->getOption('template');
+        $creationClassName = $input->getOption('class');
+        if ($altTemplate && $creationClassName) {
+            throw new \InvalidArgumentException('Cannot use --template and --class');
```
Comment (shadowhand) [inline src/Phinx/Console/Command/Create.php]: The wording here could be improved, it should clarify that the two cannot be used together:

> Cannot use --template and --class at the same time.

**Comment by shadowhand** [comment]
Looking good! Most things requested from last time appear to be fixed. :)

File: src/Phinx/Console/Command/Create.php (hunk 4 above)
Comment (RQuadling) [inline src/Phinx/Console/Command/Create.php]: Ha ha! Yes. " at the same time" really does make all the difference!

File: src/Phinx/Migration/CreationInterface.php (hunk 3 above)
Comment (RQuadling) [inline src/Phinx/Migration/CreationInterface.php]: You need to do that in a LOT of places!!! There are 30 "2014 Rob Morgan"s in the source. :-)

File: src/Phinx/Console/Command/Create.php (hunk 1 above)
Comment (RQuadling) [inline src/Phinx/Console/Command/Create.php]: I've added this as a way to report back exactly what was chosen. I would prefer to keep it.

File: src/Phinx/Migration/CreationInterface.php (hunk 3 above)
Comment (shadowhand) [inline src/Phinx/Migration/CreationInterface.php]: New files get new date, other files get date updated as they are touched. No need to global search and replace. :)

**Comment by RQuadling** [comment]
Have I got anything left to get this PR approved/merged?

**Comment by RQuadling** [comment]
Thank you.

**Pull request merged**